fix: prevent spot resource pool contention [DET-5349]

armandmcqueen opened the pull request:
## Description

There is a bug where spot requests and spot instances in AWS do not include the resource pool as part of the tags. This leads to issues where one spot resource pool creates instances, but all of the spot resource pools think they are responsible for managing the instance. This causes spot instances to get shut down because some resource pools think they are idle. This PR ensures that we include the resource pool as one of the tags to prevent this contention. 

This PR also ensures that this change doesn't lead to orphaned spot instances by deleting all spot requests and instances that do not include the `determined-resource-pool` tag when the provisioner starts up.

## Test Plan

<!---
Describe the situations in which you've tested your change, and/or a screenshot as appropriate.
Reviewers may ask questions of this test plan to ensure adequate manual coverage of changes.
-->


## Commentary (optional)

<!---
Use this section of your description to add context to the PR. Could be for particularly
tricky bits of code that could use extra scrutiny, historical context useful for reviewers, etc.
You may intentionally leave this section blank and remove the title.
--->


## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.


<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:
- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:
- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:
- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")
-->

armandmcqueen commented on master/internal/provisioner/aws_spot_legacy.go:
```diff
@@ -0,0 +1,205 @@
+	cleanup := func(ctx *actor.Context) {
+		ctx.Log().
+			WithField("codepath", "spotLegacy").
+			Infof("Starting Legacy Spot cleanup operation")
```
I will remove all of the `Infof` log lines after I have manually tested

stoksc reviewed (commented):
This pretty much lgtm, i had some nits; they're not mandatory (this is write once code) but otherwise i'd request them. mostly just curious about the retry logic/why? can we return a multierror from the cleanup and only retry then? also i think it's a good principle to not start go routines in a library-like function call but let the user do the go'ing and retrying instead.

armandmcqueen commented on master/internal/provisioner/aws_spot_legacy.go:
```diff
@@ -0,0 +1,205 @@
+		activeSpotReqs, err := c.legacyListActiveSpotInstanceRequests(ctx)
+		if err != nil {
+			ctx.Log().
+				WithField("codepath", "spotLegacy").
```
Fixed

armandmcqueen commented on master/internal/provisioner/aws_spot_legacy.go:
```diff
@@ -0,0 +1,205 @@
+
+	response, err := c.client.DescribeSpotInstanceRequests(input)
+	if err != nil {
+		return
```
Agreed. I fixed the code here as well as the place where the code was originally copied from.

armandmcqueen commented on master/internal/provisioner/aws_spot_legacy.go:
```diff
@@ -0,0 +1,205 @@
+				WithField("codepath", "spotLegacy").
+				WithError(err).
+				Debugf("cannot list active spot requests")
+		} else {
```
Fixed

armandmcqueen commented:
I addressed the comments. I also updated the log lines to what I thin they should be in production. There is one `Info` that always runs to say that legacy spot cleanup is happening and then 3 other `Info`s that will only run if there were legacy spot requests/instances that need to be cleaned up. Everything else is now `Debug` or `Error`.

The "try again after 5 minutes" logic is because the AWS Spot APIs are eventually consistent. It can take some time for newly created Spot Requests to show up in the API so if we try to just do cleanup immediately when the master starts, we may not be able to see spot requests that exit. 

It might also take time for the state of spot requests to get updated - e.g. it might say pending, when it has actually created an instance, so we try to cancel the spot requests, but we need to go back later to confirm that there isn't a spot instance running that we need to shut down (I haven't confirmed this specific case happens, but I have run into a bunch of similar problems with the Spot API being eventual consistent). More detail on how spot works can be found here: https://github.com/determined-ai/determined/blob/master/master/internal/provisioner/aws_spot.go#L30

stoksc commented on master/internal/provisioner/aws_spot_legacy.go:
```diff
@@ -0,0 +1,216 @@
+	}
+
+	// Delete spot instances associated with the active requests
+	instancesToTerminate := newSetOfStrings()
+	for _, req := range activeSpotReqs.iter() {
+		if req.InstanceID != nil {
+			instancesToTerminate.add(*req.InstanceID)
+		}
+	}
```
why not `instancesToTerminate := activeSpotReqs.instanceIds()`?

armandmcqueen replied:
fixed

stoksc commented on master/internal/provisioner/aws_spot_legacy.go:
```diff
@@ -0,0 +1,216 @@
+	input := &ec2.DescribeSpotInstanceRequestsInput{
+		Filters: []*ec2.Filter{
+			{
+				Name: aws.String(fmt.Sprintf("tag:%s", c.TagKey)),
```
Not just here but it seems like sharing this tagging logic between spot/non-spot as much as possible could've prevented this bug.

stoksc reviewed (approved):
lgtm. i get the eventually consistency bit with spots, it's much clearer now how it's written / with the comment.

stoksc commented on master/internal/provisioner/aws_spot_legacy.go:
```diff
@@ -0,0 +1,216 @@
+
+func isLegacy(request *ec2.SpotInstanceRequest) bool {
+	for _, tag := range request.Tags {
+		if *tag.Key == "determined-resource-pool" {
```
I feel like this deref is probably always safe (unless maybe the response is malformed) but i'd feel better if we checked it first since this would crash the provisioner.

armandmcqueen replied:
fixed
